framework/src/base/Assembly.C

```diff
@@ -4721,15 +4722,23 @@ Assembly::helpersRequestData()
 }
 
 void
-Assembly::havePRefinement()
+Assembly::havePRefinement(const bool user_disabled_lagrange_p_refinement)
 {
+  if (_adjusted_fe_objects_for_p_refinement)
+    // Already done did it
+    return;
+
   const Order helper_order = _mesh.hasSecondOrderElements() ? SECOND : FIRST;
   const FEType helper_type(helper_order, LAGRANGE);
-  auto setup_helpers = [&helper_type](auto & unique_helper_container,
-                                      auto & helper_container,
-                                      const unsigned int num_dimensionalities,
-                                      const bool user_added_helper_type,
-                                      auto & fe_container)
+  std::vector<FEType> lagrange_types = {helper_type};
+  if (helper_order == SECOND)
```

Comment:
and if user_disabled_lagrange_p_refinement ?

Reply:
Then the `helper_order` is still `SECOND`. The helper order is setup at Assembly construction time